### PR TITLE
Collection literals: address comments

### PR DESCRIPTION
**Pull request by cston**
Address PROTOTYPE comments.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 1]
```diff
@@ -548,18 +548,13 @@ BoundExpression convertArrayElement(BoundExpression element, TypeSymbol elementT
             }
 
             var implicitReceiver = new BoundObjectOrCollectionValuePlaceholder(syntax, isNewInstance: true, targetType) { WasCompilerGenerated = true };
-            // PROTOTYPE: When generating a List<T>, should we use the well-known
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:551]: >PROTOTYPE [](http://example.com/codeflow?start=15&length=9)

Added open issue to test plan [https://github.com/dotnet/roslyn/issues/66418](https://github.com/dotnet/roslyn/issues/66418).

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 2]
```diff
@@ -548,18 +548,13 @@ BoundExpression convertArrayElement(BoundExpression element, TypeSymbol elementT
...
-            // member List<T>.Add() rather than relying on lookup?
             var collectionInitializerAddMethodBinder = this.WithAdditionalFlags(BinderFlags.CollectionInitializerAddMethod);
             var builder = ArrayBuilder<BoundExpression>.GetInstance(node.Elements.Length);
             foreach (var element in node.Elements)
             {
                 var result = element switch
                 {
                     BoundBadExpression => element,
-                    // PROTOTYPE: Should spread elements support target type?
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:560]: >PROTOTYPE [](http://example.com/codeflow?start=23&length=9)

Added open issue to test plan [https://github.com/dotnet/roslyn/issues/66418](https://github.com/dotnet/roslyn/issues/66418).

File: src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs [hunk 3]
```diff
@@ -1173,7 +1173,6 @@ static void Main()
                     }
                 }
                 """;
-            // PROTOTYPE: Should compile and run successfully: expectedOutput: "[1, 2, 3], "
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs:1176]: >PROTOTYPE [](http://example.com/codeflow?start=15&length=9)

Natural type for collection literals is not supported, so `[x]` and `[y]` are target-typed only.

File: src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs [hunk 4]
```diff
@@ -1325,7 +1324,6 @@ static void Main()
             CompileAndVerify(new[] { source, s_collectionExtensions }, expectedOutput: "(System.Int32[]) [1, 2, 3], ");
         }
 
-        // PROTOTYPE: Test other variance cases.
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs:1328]: >PROTOTYPE [](http://example.com/codeflow?start=11&length=9)

See `TypeInference_38`, `_39`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs [hunk 5]
```diff
@@ -2087,10 +2207,6 @@ static void Main()
                     }
                 }
                 """;
-            // PROTOTYPE: Confirm we should generate ListBase<int> instances for both x and y.
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs:2090]: >PROTOTYPE [](http://example.com/codeflow?start=15&length=9)

Natural type is not supported.

File: src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs [hunk 6]
```diff
@@ -4156,7 +4272,7 @@ static void Main()
         [ConditionalTheory(typeof(CoreClrOnly))]
         [CombinatorialData]
         public void SpreadElement_01(
-            [CombinatorialValues("IEnumerable", "IEnumerable<int>", "int[]", "List<int>", "Span<int>", "ReadOnlySpan<int>")] string spreadType,
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs:4159]: >IEnumerable [](http://example.com/codeflow?start=34&length=11)

Tested separate in `SpreadElement_11`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs [hunk 7]
```diff
@@ -4584,7 +4697,6 @@ static void Main()
                 }
                 """;
             var comp = CreateCompilation(source);
-            // PROTOTYPE: Should spread elements support target type? (Should we infer IEnumerable<int>?)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs:4587]: >PROTOTYPE [](http://example.com/codeflow?start=15&length=9)

Added open issue to test plan [https://github.com/dotnet/roslyn/issues/66418](https://github.com/dotnet/roslyn/issues/66418).

File: src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs [hunk 8]
```diff
@@ -4606,7 +4718,6 @@ static string[] Append(string a, string b, bool c)
                     }
                 }
                 """;
-            // PROTOTYPE: Should spread elements support target type? (Should we infer IEnumerable<string>?)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs:4609]: >PROTOTYPE [](http://example.com/codeflow?start=15&length=9)

Added open issue to test plan [https://github.com/dotnet/roslyn/issues/66418](https://github.com/dotnet/roslyn/issues/66418).

File: src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs [hunk 9]
```diff
@@ -6303,33 +6441,5 @@ static async Task<T> F<T>(T t)
                 """;
             CompileAndVerify(new[] { source, s_collectionExtensions }, expectedOutput: "[3, 1, 2, 4], ");
         }
-
-        [ConditionalFact(typeof(CoreClrOnly), AlwaysSkip = "PROTOTYPE: 'IAsyncEnumerable<int>' does not contain a definition for 'GetAwaiter'")]
-        public void Async_03()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs:6308]: >Async_03 [](http://example.com/codeflow?start=20&length=8)

Added open issue to test plan [https://github.com/dotnet/roslyn/issues/66418](https://github.com/dotnet/roslyn/issues/66418): support `IAsyncEnumerable<T>` for spread element?

**Comment by cston** [comment]
>         // PROTOTYPE: Should this compile successfully?

Added open issue to test plan [[https://github.com/dotnet/roslyn/issues/66418](https://github.com/dotnet/roslyn/issues/66418)](https://github.com/dotnet/roslyn/issues/66418).

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs:2036 in a207dec. [](commit_id = a207decc21e28208329645f6662ecfbcc7cb4b62, deletion_comment = True)

File: src/Compilers/CSharp/Test/Syntax/Parsing/CollectionLiteralParsingTests.cs [hunk 10]
```diff
@@ -5365,8 +5364,8 @@ public void CastVersusIndexAmbiguity24_B()
     [Fact]
     public void CastVersusIndexAmbiguity24_C()
     {
-        // PROTOTYPE: This is not a great diagnostic.  Users could easily run into this and be confused.  Can we do
-        // better.  For example:
+        // This is not a great diagnostic.  Users could easily run into this and be confused.  Can we do
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Syntax/Parsing/CollectionLiteralParsingTests.cs:5367]: Tracking issue?

**Review by 333fred — APPROVED** [review]
One comment, otherwise LGTM

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 11]
```diff
@@ -3836,7 +3836,6 @@ internal uint GetValEscape(BoundExpression expr, uint scopeOfTheContainingExpres
                     return GetValEscape(switchExpr.SwitchArms.SelectAsArray(a => a.Value), scopeOfTheContainingExpression);
 
                 case BoundKind.CollectionLiteralExpression:
-                    // PROTOTYPE: Revisit if spans may be optimized to avoid heap allocation.
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:3839]: We have a test plan #66418 item tracking this now.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Syntax/Parsing/CollectionLiteralParsingTests.cs (hunk 10 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Syntax/Parsing/CollectionLiteralParsingTests.cs:5367]: This particular case involves a dictionary element which is not currently supported. I've logged an issue to improve the diagnostic for the previous example `(A)[1]`: https://github.com/dotnet/roslyn/issues/68862.

**Pull request merged**